### ISSUE TITLE
Maybe inconsistent duplicate check in `record_artifacts_as_dict`

### ISSUE THREAD

**Issue by lukpueh** [in-toto/in-toto#601]
`record_artifacts_as_dict` helper called from `runlib.in_toto_*` interface functions returns a dictionary of artifact URIs with their hashes. The hashes are generated in batches per artifact resolver and combined here:

https://github.com/in-toto/in-toto/blob/4581e4b38f2b7b9e5c6c01793b39af463805027e/in_toto/runlib.py#L182-L187

Currently, there is no URI duplicate check when combining these batches. This may not be necessary, because artifact URIs are usually distinguished by their schemes, but artifact resolver implementations may decide to arbitrarily mangle the URIs prior to returning them. Relying on a resolver to not return URIs that we already have is brittle.

At the same time, we do raise on duplicate URIs as a result of file mangling (lstrip) in `FileResolver.hash_artifacts`:

https://github.com/in-toto/in-toto/blob/4581e4b38f2b7b9e5c6c01793b39af463805027e/in_toto/resolver/_resolver.py#L116-L121

**Expected behaviour**
A user should consistently be informed, if, when using an interface function, which record artifacts, the results may not be as expected, because of duplicate URIs. IMO a warning should be enough.

**Comment by adityasaky** [comment]
I agree with adding a check at the point marked in the first snippet. I think, however, a warning may not suffice. Are there scenarios where an overwrite is fine and it's not an error? I can't think of a valid use case and I prefer to error rather than (relatively) hide away the behaviour. If the warning is missed, the metadata is less complete than the user thinks, which I'm not a fan of.

**Comment by chasen-bettinger** [comment]
@adityasaky @lukpueh Did you two want to move forward with erroring instead of warning?

**Comment by lukpueh** [comment]
@adityasaky is right, that this shouldn't go undetected, so erroring out does make sense. Not sure, which error to use though. `PrefixError` does not fit here. Adding a new custom error seems overkill, given that it's unlikely that a caller will handle it, but it probably makes sense for semantics. What do you think?

**Comment by chasen-bettinger** [comment]
@lukpueh I'm all for being explicit and clear when the option is available. Overkill? Yea, I see it. A better user experience? I think so.

**Comment by adityasaky** [comment]
Let's introduce a new exception and replace PrefixError with it when we're ready for in-toto 3.0.0, a generic error for any artifact collisions.

**Comment by PratGit1606** [comment]
I'd like to implement this. Based on the discussion above, planning to:
1. Add a new `ArtifactCollisionError` exception to `exceptions.py`
2. Add a duplicate URI check after combining resolver batches in `record_artifacts_as_dict`
3. Replace `PrefixError` with `ArtifactCollisionError` in `FileResolver.hash_artifacts` for consistency

Will open a PR shortly.
